### PR TITLE
resolve #1165

### LINKED ISSUE #1165: Treat CORES as an alias to OMP_NUM_THREADS
@akahs  found the following error while running makeflow -Tlocal:
 
libgomp: Thread creation failed: Resource temporarily unavailable
Thread creation failed: Resource temporarily unavailable

The problem is that both makeflow and the task executable (/usr/bin/convert) try to use all the cores in the machine. makeflow sets CORES=1, but convert respects OMP_NUM_THREADS. For n cores, nxn processes are created and this exhausted the maximum number of concurrent processes allowed.

This is a really nice example on how different applications respect/ignore resources limits. Note that  adding OMP_NUM_THREADS would not solve the problem in general, as the executable may ignore its value, or use some other framework (e.g., MPI).

### PR DESCRIPTION
**Pull request by charleszheng44**
Solve issue #1165, set OMP_NUM_THREADS = 1 and export it as an environment variable.

### CODE REVIEW

**Comment by dthain** [comment]
I think the point is that OMP NUM CORES should have the same value as CORES.

**Comment by charleszheng44** [comment]
Does that mean `OMP_NUM_CORES` should be set as 0 when CORES is not set?

**Comment by btovar** [comment]
No, see my comment above.

**Comment by charleszheng44** [comment]
I have updated the code.

File: makeflow/src/dag_node.c [hunk 1]
```diff
@@ -346,8 +346,15 @@ struct jx * dag_node_env_create( struct dag *d, struct dag_node *n )
 	while((key = set_next_element(d->export_vars))) {
 		char *value = dag_variable_lookup_string(key, &s);
 		if(value) {
+			if(strcmp(key, "CORES")) {
+				jx_insert(object, jx_string("OMP_NUM_THREADS"), jx_string(value));
+			}
 			jx_insert(object,jx_string(key),jx_string(value));
 			debug(D_MAKEFLOW_RUN, "export %s=%s", key, value);
+		} else {
+			if(strcmp(key, "CORES")) {
```
Comment (btovar) [inline makeflow/src/dag_node.c]: How is cores defined if it does not have a value?

File: makeflow/src/dag_node.c [hunk 2]
```diff
@@ -346,8 +346,15 @@ struct jx * dag_node_env_create( struct dag *d, struct dag_node *n )
 	while((key = set_next_element(d->export_vars))) {
 		char *value = dag_variable_lookup_string(key, &s);
 		if(value) {
+			if(strcmp(key, "CORES")) {
```
Comment (btovar) [inline makeflow/src/dag_node.c]: strcmp returns 0 if the comparison succeeds. Here you are setting OMP_NUM_THREADS to the value of each other variable, but CORES.

File: makeflow/src/parser.c [hunk 3]
```diff
@@ -149,6 +149,10 @@ void dag_close_over_environment(struct dag *d)
 			{
 				dag_variable_add_value(name, d->default_category->mf_variables, 0, value_env);
 			}
+
+			if(!value_env && !strcmp(name, RESOURCES_CORES)) {
+				dag_variable_add_value(name, d->default_category->mf_variables, 0, "1");
```
Comment (dthain) [inline makeflow/src/parser.c]: I don't think you should be changing the dag unilaterally at parse time.  If cores isn't specified, then it isn't specified.

File: makeflow/src/dag_node.c [hunk 4]
```diff
@@ -341,14 +341,19 @@ struct jx * dag_node_env_create( struct dag *d, struct dag_node *n )
 	char *key;
 
 	struct jx *object = jx_object(0);
+    
+    char *num_cores = dag_variable_lookup_string(RESOURCES_CORES, &s);
```
Comment (dthain) [inline makeflow/src/dag_node.c]: Generally speaking, check for null anytime you have a pointer returned from a function.
If it's null, then `CORES` isn't set.

**Comment by dthain** [comment]
@charleszheng44 please finish up this PR, talk to @btovar if you aren't sure how to go forward.

**Comment by charleszheng44** [comment]
1. If number of cores is not set
    + If `OMP_NUM_CORES` is not specified, we set number of cores to 1
    + else, we set number of cores to `OMP_NUM_CORES`

2. if number of cores is set
    + If `OMP_NUM_CORES` is not specified, we set `OMP_NUM_CORES` to the number of cores
    + If `OMP_NUM_CORES` is set, we do nothing

**Comment by dthain** [comment]
Wow, what happened to this PR?

**Comment by charleszheng44** [comment]
I rebase the latest master and push the change.

**Comment by dthain** [comment]
@batrick Please take a look at this PR and figure out if it safe to merge.  Discuss with @charleszheng44 as needed.

**Comment by batrick** [comment]
I helped him collapse his changes into a single commit based on master. If
should merge cleanly now. Charles is checking if it is still does what it's
supposed to do.

**Pull request merged**